text/0000-reactive-assignments.md

```diff
@@ -0,0 +1,946 @@
+
+> 🐃 We could match the arguments to `new CustomEvent(name, opts)` instead — where `detail` is one of the options passed to the constructor alongside things like `bubbles` and `cancelable`
+
+As with lifecycle functions, the `dispatch` is bound to the component because of when `createEventDispatcher` is called.
+
+Listening to events is currently done like so:
+
+```html
+<button on:click="doSomething()">click me</button>
+```
+
+This effectively becomes `{() => this.doSomething()}`. The shorthand is nice, but it is slightly confusing (the context is implicit, and the right-hand side must be a CallExpression which limits flexibility, although there's an [issue for that](https://github.com/sveltejs/svelte/issues/1766)). In the new model, there are no longer any problems around `this`, so it probably makes sense to allow arbitrary expressions instead.
+
+
+### Bindings
+
+Element bindings are a convenient shorthand, which I believe we should preserve:
+
+```html
+<input bind:value=name>
+
+<!-- equivalent (Svelte 2) -->
+<input value={name} on:input="set({ name: this.value })">
+
+<!-- equivalent (this proposal) -->
+<input value={name} on:input="{e => name = e.target.value}">
+```
+
+Component bindings are less clear-cut. In the past they've been a source of confusion for the same reason that the lifecycle hook dance causes confusion upon app initialisation. It's possible that the new lifecycle concepts will eliminate those headaches, but this requires further exploration.
+
+The reason we can *consider* removing component bindings is that it's now trivial to pass state-altering functions down to components:
+
+```html
+<script>
+  import Mousecatcher from './Mousecatcher.html';
+
+  let mouse;
+
+  function setPosition(x, y) {
+    mouse = { x, y };
+  }
+</script>
+
+<Mousecatcher {setPosition}>
+  the mouse is at {mouse.x},{mouse.y}
+</Mousecatcher>
+```
+
+
+### Component API
+
+Under this proposal, there is no longer a `this` with state and methods *inside* a component. But there needs to be a way to interact with the component from the outside world.
+
+Instantiating a top-level component probably needn't change (except 🐃 maybe changing `data` to `props`):
+
+```js
+import App from './App.html';
+
+const app = new App({
+  target: document.querySelector('body'),
+  props: {
+    name: 'world'
+  }
+});
+```
+
+Exported properties can be exposed as accessors:
+
+```js
+app.name; // world
+app.name = 'everybody'; triggers a (sync?) update
+```
+
+This creates consistent behaviour between Svelte components that are compiled to custom elements, and those that are not, while also making it easy to understand the component's contract.
+
+Of the five **built-in methods** that currently comprise the [component API](https://svelte.technology/guide#component-api) — `get`, `set`, `fire`, `on` and `destroy` — we no longer need the first three. `on` and `destroy` are still necessary.
```

Comment:
Do we want to have some naming/prefixing convention for built-in methods (or data, like the current `options` or `root`)? This is a little less easy to type, but it does avoid issues in the future if we want to add something and we need to figure out whether that's a breaking changing change because it might conflict with user-defined items. We can tell users 'don't name anything starting with a `$`' but we can't say 'don't name anything that sounds like we might want to use it later potentially'.

Reply:
Not a bad idea. Vue does `$` prefixes. As long as there's no point of confusion with Store, that's probably a good move

Reply:
I don't understand why `fire` and `on` are removed?

A method's sole purpose may be to emit an event, and rather than pass the Component a `props.onFoobar` , something else can listen to Component for that sequence

```js
const greet = new Greeter();
const stutter = str => str.charAt(0).repeat(3) + str;

greet.name = 'Bob';
greet.name = 'George';

greet.on('pretty-girl', stutter);
```

Reply:
^ My concerns erased. Solved in chat.